core/store/src/adapter/chain_store.rs

```diff
@@ -390,6 +396,95 @@ impl ChainStoreAdapter {
     }
 }
 
+pub struct ChainStoreUpdateAdapter<'a> {
+    store_update: StoreUpdateHolder<'a>,
+}
+
+impl Into<StoreUpdate> for ChainStoreUpdateAdapter<'static> {
+    fn into(self) -> StoreUpdate {
+        self.store_update.into()
+    }
+}
+
+impl ChainStoreUpdateAdapter<'static> {
+    pub fn commit(self) -> io::Result<()> {
+        let store_update: StoreUpdate = self.into();
+        store_update.commit()
+    }
+}
+
+impl<'a> StoreUpdateAdapter for ChainStoreUpdateAdapter<'a> {
+    fn store_update(&mut self) -> &mut StoreUpdate {
+        &mut self.store_update
+    }
+}
+
+impl<'a> ChainStoreUpdateAdapter<'a> {
+    pub fn new(store_update: &'a mut StoreUpdate) -> Self {
+        Self { store_update: StoreUpdateHolder::Reference(store_update) }
+    }
+
+    /// USE THIS FUNCTION WITH CARE; proceed only if you know what you're doing.
+    /// Typically while saving the block header we would also like to update
+    /// block_header_hashes_by_height and update block_merkle_tree
+    pub fn set_block_header_only(&mut self, header: &BlockHeader) {
+        self.store_update.insert_ser(DBCol::BlockHeader, header.hash().as_ref(), header).unwrap();
+    }
+
+    /// USE THIS FUNCTION WITH CARE; proceed only if you know what you're doing.
+    /// Typically block_header_hashes_by_height is saved while saving the block header
+    pub fn set_block_header_hashes_by_height(
+        &mut self,
+        height: BlockHeight,
+        hash_set: &HashSet<CryptoHash>,
+    ) {
+        self.store_update
+            .set_ser(DBCol::HeaderHashesByHeight, &index_to_bytes(height), hash_set)
+            .unwrap();
+    }
+
+    /// USE THIS FUNCTION WITH CARE; proceed only if you know what you're doing.
+    /// Typically block_merkle_tree is saved while saving the block header
+    pub fn set_block_merkle_tree(
+        &mut self,
+        block_hash: &CryptoHash,
+        block_merkle_tree: &PartialMerkleTree,
+    ) {
+        self.store_update
+            .set_ser(DBCol::BlockMerkleTree, block_hash.as_ref(), block_merkle_tree)
+            .unwrap();
+    }
+
+    pub fn set_block_ordinal(&mut self, block_ordinal: NumBlocks, block_hash: &CryptoHash) {
+        self.store_update
+            .set_ser(DBCol::BlockOrdinal, &index_to_bytes(block_ordinal), block_hash)
+            .unwrap();
+    }
+
+    pub fn set_block_height(&mut self, hash: &CryptoHash, height: BlockHeight) {
+        self.store_update
+            .set_ser(DBCol::BlockHeight, &borsh::to_vec(&height).unwrap(), hash)
+            .unwrap();
+    }
+
+    pub fn set_header_head(&mut self, header_head: &Tip) {
+        self.store_update.set_ser(DBCol::BlockMisc, HEADER_HEAD_KEY, header_head).unwrap();
+    }
+
+    pub fn set_final_head(&mut self, final_head: &Tip) {
+        self.store_update.set_ser(DBCol::BlockMisc, FINAL_HEAD_KEY, final_head).unwrap();
+    }
+
+    /// This function is normally clubbed with set_block_header_only
```

Comment:
But then we can just unite both methods.

Reply:
Yes, I was thinking of doing that but thought of leaving the primitives. I had plans in the future to integrate functions like these into a helper module like, `update_block` or similar, but thought of leaving them as is for now.

I can follow up in a different PR